Introduce liblibctest for unikraft

flpostolache opened the pull request:
This is a draft PR that introduces the liblibctest library for Unikraft. Please note that this isn't a final version. Suggestions are welcome.
To run the tests, you need to apply this commit over the unikraft repo to enable downloading a lib as a github repo [1]. Also, if you want to test the lib on musl, you need to disable the myself test suite. This PR enables the disabling of those tests [2].

[1]: https://github.com/unikraft/unikraft/pull/553
[2]: https://github.com/unikraft/unikraft/pull/602

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Initial commit

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Convert functional, musl and regression tests to work with uktest

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Order patches and start grouping sources

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

razvand commented on Config.uk:
```diff
@@ -0,0 +1,13 @@
+menuconfig LIBLIBCTEST
+	bool "Libc-test Unikraft library"
+	default n
```
```suggestion
	default n
	select LIBUKTEST
```

flpostolache replied:
Solved

razvand reviewed (changes requested):
Add dependency on `LIBUKTEST`.

DeliaPavel commented on patches/0008-Modify-env-source-to-work-with-uktest.patch:
```diff
@@ -0,0 +1,93 @@
+-	if ((r=setenv("","",0)) != -1 || errno != EINVAL)
+-		t_error("setenv(\"\",\"\",0): %d, errno: %d (%s), wanted -1, %d (EINVAL)\n", r, errno, strerror(errno), EINVAL);
+-	return t_status;
++	UK_TEST_ASSERT(((r=setenv("","",0)) != -1 || errno != EINVAL));	
```
```suggestion
+	UK_TEST_ASSERT(!((r=setenv("","",0)) != -1 || errno != EINVAL));	
```

flpostolache replied:
Solved

flpostolache committed:
Group tests and sources based on functionality

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Add new access test

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

DeliaPavel commented on patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch:
```diff
@@ -0,0 +1,80 @@
+ 	pthread_mutexattr_t mtx_a;
+ 	pthread_mutex_t mtx;
+-
++	for(int pshared=0; pshared<=1; pshared++) for(int pi=0; pi<=1; pi++){
```
Is there any reason for these `for` statements?

flpostolache replied:
No. I removed them.

DeliaPavel commented on patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch:
```diff
@@ -0,0 +1,80 @@
+-	if (r && r != PTHREAD_BARRIER_SERIAL_THREAD)
+-		t_error("pthread_barrier_wait failed: got %d \"%s\", wanted either 0 or %d\n",
+-			r, strerror(r), PTHREAD_BARRIER_SERIAL_THREAD);
++	if(!(r && r != PTHREAD_BARRIER_SERIAL_THREAD))
```
Maybe this if statement should keep the previous condition. I am not sure how it should be, but to me, it makes more sense the other way around.
```suggestion
+	if(r && r != PTHREAD_BARRIER_SERIAL_THREAD)
```

flpostolache replied:
Changed it.

DeliaPavel commented on patches/0096-Modify-strptime-source-to-work-with-uktest.patch:
```diff
@@ -0,0 +1,124 @@
+-	checkStrptimeTz("-06", -6, 0);
+-
+-	return t_status;
++	UK_TEST_EXPECT_ZERO(checkStrptime("1856-07-10", "%F", &tm4));
```
These tests fail and, judging by the comment, they might be applicable only for glibc.

flpostolache replied:
Yes, they are currently not working in `musl`. It is not a problem of Unikraft. @razvand Should they be commented? What do you think?

razvand replied:
@maniatro111 , I think they are OK at this point. At some point we would probably do a selection based on libc-type. But I wouldn't make it part of this release.

Having things working (i.e. all tests running, NOT passing) is fine. There shouldn't be crashes when running the tests. Other than that, it's fine.

DeliaPavel commented on Makefile.uk:
```diff
@@ -0,0 +1,319 @@
+	LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_memcpy.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strchr.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strcspn.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strstr.c
```
As far as I could tell, there is no strstr implementation in nolibc so maybe this should be moved in the ifeq as well.

flpostolache replied:
Moved it.

flpostolache committed:
Removed dependency on stdio_impl.h

This commit removes dependency on stdio_impl.h and defaults all
test cases to n.

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Modify sources to print more detailed outputs

Need to renumber them.

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

kubanrob commented:
I had a problem where I wanted to exclude crashing tests from the config. For that I need to check which config option includes that test, looking it up in the Makefiles.
It would be useful to display some kind of name related to the group names displayed in menuconfig. Maybe there could be one testsuite per config option instead of adding all to the same.

flpostolache committed:
Separate tests into multiple testsuites

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Separate test groups into multiple testsuites

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Modify patch files to print more info

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Make some sources print mor info

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Add newlines at the end of some sources

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache committed:
Add some configuration details to README.md file

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

flpostolache commented:
@kubanrob I modified the test structure. Now, each test case displays the group name from which is a part of and the name of the source file. If you want to disable the whole group, you can now get the name from the test case label that crashes and simply disable it from the menuconfig. If you want to only disable that specific source, you can get its name form the test case label and comment that line in the Makefile. I think in the future, an option to disable a specific source will be useful.

DeliaPavel reviewed (approved):
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

kubanrob reviewed (approved):
Thank you @maniatro111

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>

razvand reviewed (approved):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

flpostolache committed:
Separate git clone and patch operations

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>
